python/pywatchman/bser.c

```diff
@@ -1084,9 +1087,87 @@ static PyObject *bser_loads(PyObject *self, PyObject *args, PyObject *kw)
   return bser_loads_recursive(&data, end, &ctx);
 }
 
+static PyObject *bser_load(PyObject *self, PyObject *args, PyObject *kw)
+{
+  size_t rc = 0;
+  FILE *fp = NULL;
+  char header[SNIFF_BUFFER_SIZE];
+  const char *data = NULL;
+  char *newdata = NULL;
+  const char *end;
+  int64_t expected_len, total_len;
+  off_t position;
+  PyObject *fp_obj = NULL;
+  PyObject *mutable_obj = NULL;
+  const char *value_encoding = NULL;
+  const char *value_errors = NULL;
+  unser_ctx_t ctx = {1, 0};
+
+  PyObject *ret = NULL;
+
+  static char *kw_list[] = {"fp", "mutable", "value_encoding", "value_errors",
+                            NULL};
+
+  if (!PyArg_ParseTupleAndKeywords(args, kw, "O|Ozz:loads", kw_list, &fp_obj,
+                                   &mutable_obj, &value_encoding,
+                                   &value_errors)) {
+    goto out;
+  }
+
+  if (mutable_obj) {
+    ctx.mutable = PyObject_IsTrue(mutable_obj) > 0 ? 1 : 0;
+  }
+  ctx.value_encoding = value_encoding;
+  if (value_encoding == NULL) {
+    ctx.value_errors = NULL;
+  } else if (value_errors == NULL) {
+    ctx.value_errors = "strict";
+  } else {
+    ctx.value_errors = value_errors;
+  }
+
+  if (!PyFile_Check(fp_obj)) {
+    PyErr_SetString(PyExc_ValueError, "invalid file object");
+    goto out;
+  }
+  fp = PyFile_AsFile(fp_obj);
+
+  rc = fread(header, 1, SNIFF_BUFFER_SIZE, fp);
+  if (rc != SNIFF_BUFFER_SIZE) {
+    PyErr_SetString(PyExc_ValueError, "bser data ended early");
+    goto out;
+  }
+  data = header;
+  if (!_pdu_info_helper(data, data + SNIFF_BUFFER_SIZE, &ctx.bser_version,
+                        &ctx.bser_capabilities, &expected_len, &position)) {
+    goto out;
+  }
+  total_len = expected_len + position;
+  newdata = (char *)malloc(total_len);
+  if (newdata == NULL) {
+    goto out;
+  }
+  memcpy(newdata, header, SNIFF_BUFFER_SIZE);
+  rc = fread(newdata + SNIFF_BUFFER_SIZE, 1, total_len - SNIFF_BUFFER_SIZE, fp);
+  if (rc != total_len - SNIFF_BUFFER_SIZE) {
+    PyErr_SetString(PyExc_ValueError, "bser data ended early");
```

Comment:
I agree that I don't think read will get everything in one shot, but my understanding is that fread is supposed to have that guarantee.  The only reason we would get less than the full amount is if there's an EOF or a read error